internal/pkg/stream/stream.go

```diff
@@ -0,0 +1,52 @@
+// Package stream implements streamers that publish AWS events periodically.
+// A streamer fetches AWS events periodically and notifies subscribed channels of them.
+package stream
+
+import (
+	"context"
+	"time"
+)
+
+// Fetcher fetches events, updates its internal state with new events and returns the next time
+// the Fetch call should be attempted. On failure, Fetch returns an error.
+type Fetcher interface {
+	Fetch() (next time.Time, err error)
+}
+
+// Notifier notifies all of its subscribers of any new event updates.
+type Notifier interface {
+	Notify()
+}
+
+// FetchNotifier is the interface that groups the Fetch and a Notify methods.
+type FetchNotifier interface {
+	Fetcher
+	Notifier
+}
+
+// Stream streams event updates by calling Fetch followed with Notify until the context is canceled or Fetch errors.
+// Once the context is canceled, a best effort Fetch and Notify is called one last time.
+func Stream(ctx context.Context, fn FetchNotifier) error {
+	var next time.Time
+	var err error
+	for {
+		var fetchDelay time.Duration // By default there is no delay.
+		if now := time.Now(); next.After(now) {
+			fetchDelay = next.Sub(now)
+		}
+
+		select {
+		case <-ctx.Done():
+			// The parent context is canceled. Try Fetch and Notify one last time and exit successfully.
+			_, _ = fn.Fetch()
+			fn.Notify()
```

Comment:
Can we combine fetch and notify? Like after the reverse we just output it.

Reply:
I didn't combine the two on purpose because `Notify` doesn't need to be always called by a Streamer and a client might want the flexibility on when to get notified. Here is the blurb from the ["Observer" pattern](https://www.amazon.com/Design-Patterns-Elements-Reusable-Object-Oriented/dp/0201633612):  
> (b) Make clients responsible for calling `Notify` at the right time. The advantage here is that the client can wait to trigger the update until a series of state changes has been made, thereby avoiding needless intermediate updates. The disadvantage is that clients have an added responsibility to trigger the update. That makes errors more likely, since clients might forget to call `Notify`.